File: android/app/src/main/java/com/tacmap/map/TerrainHeatmapService.kt
package com.tacmap.map

import android.graphics.Bitmap
import android.graphics.Color
import com.tacmap.calibration.Wgs84Bounds
import com.tacmap.settings.OpsecSettings
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.util.Locale

/**
 * Auto terrain heatmap for a WGS84 region. Samples a grid of elevations
 * from Open-Meteo's Copernicus DEM and colours blue (low) to red (high).
 * No user-uploaded file needed, unlike the competitor's manual import.
 *
 * Returned bitmap is upscaled + alpha-blended for a smooth semi-transparent
 * overlay when stretched across the region.
 */
class TerrainHeatmapService {

    @Serializable
    private data class Response(val elevation: List<Double> = emptyList())

    private val json = Json { ignoreUnknownKeys = true }

    /** Sample [grid]x[grid] elevations over [bounds] and return a coloured,
     *  upscaled bitmap, or null on failure. */
    suspend fun generate(bounds: Wgs84Bounds, grid: Int = 24): Bitmap? = withContext(Dispatchers.IO) {
        // OPSEC: sampling the DEM sends coordinates to Open-Meteo,
        // only proceed if user opted into online lookups.
        if (OpsecSettings.shared?.onlineLookups?.value != true) return@withContext null
        val south = bounds.southwest.latitude
        val north = bounds.northeast.latitude
        val west = bounds.southwest.longitude
        val east = bounds.northeast.longitude
        if (grid !in 2..32 || !south.isFinite() || !north.isFinite() ||
            !west.isFinite() || !east.isFinite() || south !in -90.0..90.0 ||
            north !in -90.0..90.0 || west !in -180.0..180.0 || east !in -180.0..180.0
        ) return@withContext null

        // Row-major grid; row 0 = north edge, col 0 = west edge (GroundOverlay's
        // image origin is the NW corner).
        val lat = DoubleArray(grid * grid)
        val lon = DoubleArray(grid * grid)
        for (r in 0 until grid) {
            val y = north - (north - south) * r / (grid - 1)
            for (c in 0 until grid) {
                val idx = r * grid + c
                lat[idx] = y
                lon[idx] = west + (east - west) * c / (grid - 1)
            }
        }

        val elev = DoubleArray(grid * grid) { Double.NaN }
        var i = 0
        while (i < elev.size) {
            // If user panned away or toggled heatmap off, bail early
            // instead of finishing every batch.
            ensureActive()
            val end = minOf(i + 100, elev.size)   // Open-Meteo: <=100 points/request
            // Coarsen to ~11 m before egress: more precision is useless for
            // this 24x24 visualisation and needlessly fingerprints the AO.
            val latStr = (i until end).joinToString(",") { String.format(Locale.US, "%.4f", lat[it]) }
            val lonStr = (i until end).joinToString(",") { String.format(Locale.US, "%.4f", lon[it]) }
            val body = fetch(
                "https://api.open-meteo.com/v1/elevation?latitude=$latStr&longitude=$lonStr"
            ) ?: return@withContext null
            val parsed = runCatching { json.decodeFromString<Response>(body) }.getOrNull()
                ?: return@withContext null
            for (k in parsed.elevation.indices) if (i + k < elev.size) elev[i + k] = parsed.elevation[k]
            i = end
        }

        val valid = elev.filter { !it.isNaN() }
        if (valid.isEmpty()) return@withContext null
        val min = valid.min()
        val max = valid.max()
        val range = (max - min).takeIf { it > 1e-6 } ?: 1.0

        val small = Bitmap.createBitmap(grid, grid, Bitmap.Config.ARGB_8888)
        for (r in 0 until grid) {
            for (c in 0 until grid) {
                val e = elev[r * grid + c]
                small.setPixel(c, r, if (e.isNaN()) Color.TRANSPARENT else colorFor((e - min) / range))
            }
        }
        // Upscale with bilinear filtering for a smooth gradient overlay.
        val scaled = Bitmap.createScaledBitmap(small, 256, 256, true)
        small.recycle()
        scaled
    }

    /** Blue (t=0, low) → cyan → green → yellow → red (t=1, high), ~45% alpha. */
    private fun colorFor(t: Double): Int {
        val clamped = t.coerceIn(0.0, 1.0)
        val hue = ((1.0 - clamped) * 240.0).toFloat()  // 240=blue .. 0=red
        val rgb = Color.HSVToColor(floatArrayOf(hue, 0.85f, 0.95f)) and 0x00FFFFFF
        return (0x73 shl 24) or rgb   // 0x73 ≈ 45% alpha
    }

    private suspend fun fetch(url: String): String? = boundedHttpsGet(url, MAX_RESPONSE_BYTES)

    private companion object { const val MAX_RESPONSE_BYTES = 1024 * 1024 }
}

File: ios/TacticalMaps/Map/TerrainHeatmapService.swift
import Foundation
import UIKit
import MapKit

/// Auto terrain heatmap - samples a grid of elevations from Open-Meteo's
/// Copernicus DEM (same source as elevation readout) and colours them
/// blue (low) to red (high). No user-uploaded file needed, unlike the
/// competitor's manual elevation-file import. Swift mirror of Android
/// TerrainHeatmapService.
actor TerrainHeatmapService {

    private struct Response: Decodable { let elevation: [Double] }

    /// Sample a gridxgrid DEM over region and return a coloured image.
    func generate(region: MKCoordinateRegion, grid: Int = 24) async -> UIImage? {
        // OPSEC: the heat-map samples the region's DEM from a third party
        // (Open-Meteo), transmitting the coordinates. Opt-in only.
        guard OpsecSettings.shared.onlineLookups else { return nil }
        guard grid >= 2 else { return nil }
        let north = region.center.latitude + region.span.latitudeDelta / 2
        let south = region.center.latitude - region.span.latitudeDelta / 2
        let west = region.center.longitude - region.span.longitudeDelta / 2
        let east = region.center.longitude + region.span.longitudeDelta / 2

        var lat = [Double](repeating: 0, count: grid * grid)
        var lon = [Double](repeating: 0, count: grid * grid)
        for r in 0..<grid {           // row 0 = north edge
            let y = north - (north - south) * Double(r) / Double(grid - 1)
            for c in 0..<grid {       // col 0 = west edge
                let idx = r * grid + c
                lat[idx] = y
                lon[idx] = west + (east - west) * Double(c) / Double(grid - 1)
            }
        }

        var elev = [Double?](repeating: nil, count: grid * grid)
        var i = 0
        while i < elev.count {
            guard OpsecSettings.shared.onlineLookups, !Task.isCancelled else { return nil }
            let end = min(i + 100, elev.count)   // Open-Meteo: <=100 points/request
            let latStr = (i..<end).map { String(lat[$0]) }.joined(separator: ",")
            let lonStr = (i..<end).map { String(lon[$0]) }.joined(separator: ",")
            guard var comps = URLComponents(string: "https://api.open-meteo.com/v1/elevation") else { return nil }
            comps.queryItems = [
                URLQueryItem(name: "latitude", value: latStr),
                URLQueryItem(name: "longitude", value: lonStr)
            ]
            guard let url = comps.url else { return nil }
            var req = URLRequest(url: url); req.timeoutInterval = 8
            do {
                req.cachePolicy = .reloadIgnoringLocalAndRemoteCacheData
                let (data, response) = try await NetworkSession.data(for: req, maximumBytes: 256 * 1024)
                guard OpsecSettings.shared.onlineLookups, !Task.isCancelled,
                      let http = response as? HTTPURLResponse,
                      (200...299).contains(http.statusCode) else { return nil }
                let decoded = try JSONDecoder().decode(Response.self, from: data)
                for k in decoded.elevation.indices where i + k < elev.count { elev[i + k] = decoded.elevation[k] }
            } catch {
                return nil
            }
            i = end
        }

        let known = elev.compactMap { $0 }
        guard let lo = known.min(), let hi = known.max() else { return nil }
        let range = (hi - lo) > 1e-6 ? (hi - lo) : 1.0

        return renderImage(grid: grid, elev: elev, lo: lo, range: range)
    }

    /// Render into a grid x grid image. UIKit/MapKit handles the upscaling.
    private func renderImage(grid: Int, elev: [Double?], lo: Double, range: Double) -> UIImage? {
        let fmt = UIGraphicsImageRendererFormat()
        fmt.scale = 1
        fmt.opaque = false
        let renderer = UIGraphicsImageRenderer(size: CGSize(width: grid, height: grid), format: fmt)
        return renderer.image { rctx in
            let ctx = rctx.cgContext
            for r in 0..<grid {
                for c in 0..<grid {
                    guard let e = elev[r * grid + c] else { continue }
                    let t = max(0.0, min(1.0, (e - lo) / range))
                    // Blue (low) -> red (high), ~45% alpha.
                    let hue = CGFloat((1.0 - t) * 240.0 / 360.0)
                    UIColor(hue: hue, saturation: 0.85, brightness: 0.95, alpha: 0.45).setFill()
                    ctx.fill(CGRect(x: c, y: r, width: 1, height: 1))
                }
            }
        }
    }
}
